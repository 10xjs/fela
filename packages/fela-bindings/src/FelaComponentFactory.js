/* @flow */
//import assignStyle from 'css-in-js-utils/lib/assignStyle'
import arrayReduce from 'fast-loops/lib/arrayReduce'

function assignStyle(a, b) {
  return Object.assign({}, a, b)
}
function combineRules(...rules: Array<Function>): Function {
  return (props, renderer) =>
    arrayReduce(
      rules,
      (style, rule) =>
        assignStyle(
          style,
          typeof rule === 'function' ? rule(props, renderer) : rule
        ),
      {}
    )
}

function resolveRule(style, extend) {
  if (extend) {
    return combineRules(style, extend)
  }

  if (typeof style === 'object') {
    return () => style
  }

  return style
}

export default function FelaComponentFactory(
  createElement: Function,
  FelaTheme: Function,
  contextTypes?: Object
): Function {
  function FelaComponent(
    { children, as = 'div', style, extend, ...otherProps },
    { renderer }
  ) {
    return createElement(FelaTheme, undefined, theme => {
      const rule = resolveRule(style, extend)

      const className = renderer.renderRule(rule, { theme, ...otherProps })

      if (children instanceof Function) {
        return children({
          className,
          theme,
        })
      }

      return createElement(as, { className }, children)
    })
  }

  if (contextTypes) {
    FelaComponent.contextTypes = contextTypes
  }

  return FelaComponent
}
