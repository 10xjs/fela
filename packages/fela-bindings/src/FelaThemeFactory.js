/* @flow */
import deprecate from './_deprecate'

export default function FelaThemeFactory(
  createElement: Function,
  ThemeContext: any
): any {
  function FelaTheme({ children, render }) {
    // TODO: remove with 11.0.0
    deprecate(
      render !== undefined,
      'The `render` prop in FelaTheme is deprecated. It will be removed in react-fela@11.0.0.\nPlease always use `children` instead. See http://fela.js.org/docs/api/bindings/fela-theme'
    )

    const renderFn = children || render

    return createElement(ThemeContext.Consumer, undefined, renderFn)
  }

  return FelaTheme
}
