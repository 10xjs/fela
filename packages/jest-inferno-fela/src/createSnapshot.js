import { render } from 'inferno'
import { createElement } from 'inferno-create-element'
import { createRenderer } from 'fela'
import { Provider, ThemeProvider } from 'inferno-fela'

import { createSnapshotFactory } from 'jest-fela-bindings'

export default createSnapshotFactory(
  createElement,
  render,
  createRenderer(),
  Provider,
  ThemeProvider
)
