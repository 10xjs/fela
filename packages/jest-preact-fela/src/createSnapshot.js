import { createElement, render } from 'preact'
import { createRenderer } from 'fela'
import { RendererProvider, ThemeProvider } from 'preact-fela'

import { createSnapshotFactory } from 'jest-fela-bindings'

export default createSnapshotFactory(
  createElement,
  render,
  createRenderer(),
  RendererProvider,
  ThemeProvider
)
