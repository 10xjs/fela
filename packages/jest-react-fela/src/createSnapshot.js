import { createElement } from 'react'
import { render } from 'react-dom'
import { createRenderer } from 'fela'
import { Provider, ThemeProvider } from 'react-fela'

import { createSnapshotFactory } from 'jest-fela-bindings'

export default createSnapshotFactory(
  createElement,
  render,
  createRenderer(),
  Provider,
  ThemeProvider
)
