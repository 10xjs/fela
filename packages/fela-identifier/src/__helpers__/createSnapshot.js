import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { render } from 'react-dom'
import { css } from 'js-beautify'

import RendererProviderFactory from '../../../fela-bindings/src/RendererProviderFactory'
import ThemeProviderFactory from '../../../fela-bindings/src/ThemeProviderFactory'
import { THEME_CHANNEL } from '../../../fela-bindings/src/themeChannel'

import renderToString from '../../../fela-tools/src/renderToString'
import createRenderer from '../../../fela/src/createRenderer'

const RendererProvider = RendererProviderFactory(
  Component,
  children => children,
  {
    childContextTypes: { renderer: PropTypes.object },
  }
)

const BaseThemeProvider = ThemeProviderFactory(
  Component,
  children => children,
  {
    propTypes: {
      theme: PropTypes.object.isRequired,
      overwrite: PropTypes.bool,
    },
    childContextTypes: {
      [THEME_CHANNEL]: PropTypes.object,
    },
    contextTypes: {
      [THEME_CHANNEL]: PropTypes.object,
    },
    defaultProps: {
      overwrite: false,
    },
  }
)

export default function createSnapshot(
  component,
  theme = {},
  renderer = createRenderer(),
  Provider = RendererProvider,
  ThemeProvider = BaseThemeProvider
) {
  const div = document.createElement('div')

  render(
    <Provider renderer={renderer}>
      <ThemeProvider theme={theme}>{component}</ThemeProvider>
    </Provider>,
    div
  )

  return {
    html: div.innerHTML,
    css: css(renderToString(renderer)),
  }
}
