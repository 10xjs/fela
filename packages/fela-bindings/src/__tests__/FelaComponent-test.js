import 'raf/polyfill'
import React, { Component, createElement } from 'react'
import PropTypes from 'prop-types'

import FelaThemeFactory from '../FelaThemeFactory'
import FelaComponentFactory from '../FelaComponentFactory'
import { THEME_CHANNEL } from '../themeChannel'

import createSnapshot from '../__helpers__/createSnapshot'

const FelaTheme = FelaThemeFactory(Component, {
  [THEME_CHANNEL]: PropTypes.object,
})

const FelaComponent = FelaComponentFactory(createElement, FelaTheme, {
  [THEME_CHANNEL]: PropTypes.object,
  renderer: PropTypes.object,
})

describe('Using the FelaComponent component', () => {
  it('should correctly render a fela rule', () => {
    expect(
      createSnapshot(
        <FelaComponent
          style={{
            fontSize: '12px',
            color: 'red',
          }}>
          {({ className }) => (
            <div className={className}>I am red and written in 12px.</div>
          )}
        </FelaComponent>
      )
    ).toMatchSnapshot()
  })

  it('should accept an array as style', () => {
    expect(
      createSnapshot(
        <FelaComponent
          style={[
            {
              fontSize: '12px',
              color: 'red',
            },
            {
              backgroundColor: 'black',
            },
          ]}>
          {({ className }) => (
            <div className={className}>
              I am red on black and written in 12px.
            </div>
          )}
        </FelaComponent>
      )
    ).toMatchSnapshot()
  })

  it('should accept an array as style', () => {
    expect(
      createSnapshot(
        <FelaComponent
          bgColor="blue"
          style={[
            {
              fontSize: '12px',
              color: 'red',
            },
            ({ bgColor }) => ({
              backgroundColor: bgColor,
            }),
          ]}>
          {({ className }) => (
            <div className={className}>
              I am red on blue and written in 12px.
            </div>
          )}
        </FelaComponent>
      )
    ).toMatchSnapshot()
  })

  it('should correctly pass the theme and other props to functional style', () => {
    const rule = ({ theme, bgc }) => ({
      fontSize: theme.fontSize,
      backgroundColor: bgc || 'red',
    })

    expect(
      createSnapshot(
        <FelaComponent style={rule} bgc="blue">
          {({ className, theme }) => (
            <div className={className}>
              I am red and written in {theme.fontSize}.
            </div>
          )}
        </FelaComponent>,
        { fontSize: '15px' }
      )
    ).toMatchSnapshot()
  })

  it('should default to a div', () => {
    expect(
      createSnapshot(
        <FelaComponent
          style={{
            fontSize: '12px',
            color: 'red',
          }}
        />
      )
    ).toMatchSnapshot()
  })

  it('should render children in default mode', () => {
    expect(
      createSnapshot(
        <FelaComponent
          style={{
            fontSize: '12px',
            color: 'red',
          }}>
          <span>Hello World</span>
        </FelaComponent>
      )
    ).toMatchSnapshot()
  })

  it('should accept a string primitive type via as-prop', () => {
    expect(
      createSnapshot(
        <FelaComponent
          as="span"
          style={{
            fontSize: '12px',
            color: 'red',
          }}
        />
      )
    ).toMatchSnapshot()
  })

  it('should render children using the correct as-prop', () => {
    expect(
      createSnapshot(
        <FelaComponent
          as="h1"
          style={{
            fontSize: '12px',
            color: 'red',
          }}>
          Hello World
        </FelaComponent>
      )
    ).toMatchSnapshot()
  })

  it('should compose extension rules', () => {
    const Button = props => (
      <FelaComponent style={{ fontSize: '12px', color: 'red' }} {...props} />
    )

    expect(
      createSnapshot(
        <Button extend={{ backgroundColor: 'blue' }} as="button">
          Hello World
        </Button>
      )
    ).toMatchSnapshot()
  })

  it('should compose extension rules', () => {
    const rule = ({ color }) => ({
      fontSize: '12px',
      color,
    })

    const Button = ({ children, ...props }) => (
      <FelaComponent style={rule} {...props}>
        {({ className }) => <button className={className}>{children}</button>}
      </FelaComponent>
    )

    expect(
      createSnapshot(
        <Button extend={{ backgroundColor: 'blue' }} color="black">
          Hello World
        </Button>
      )
    ).toMatchSnapshot()
  })

  it('should compose nested extensions', () => {
    const rule = ({ color }) => ({
      fontSize: '12px',
      color,
    })

    const Button = ({ children, ...props }) => (
      <FelaComponent style={rule} {...props}>
        {({ className }) => <button className={className}>{children}</button>}
      </FelaComponent>
    )

    const ExtendedButton = ({ children, extend = {}, padding, ...props }) => (
      <Button
        {...props}
        extend={[{ border: '1px solid grey', padding }, extend]}>
        {children}
      </Button>
    )

    expect(
      createSnapshot(
        <ExtendedButton
          extend={{ backgroundColor: 'blue' }}
          color="black"
          padding="10px">
          Hello World
        </ExtendedButton>
      )
    ).toMatchSnapshot()
  })
})
