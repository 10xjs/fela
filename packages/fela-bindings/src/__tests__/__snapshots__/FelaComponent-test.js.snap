// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Using the FelaComponent component should accept a string primitive type via as-prop 1`] = `
Object {
  "css": ".a {
    font-size: 12px
}

.b {
    color: red
}",
  "html": "<span class=\\"a b\\"></span>",
}
`;

exports[`Using the FelaComponent component should compose extension rules 1`] = `
Object {
  "css": ".a {
    font-size: 12px
}

.b {
    color: red
}

.c {
    background-color: blue
}",
  "html": "<button class=\\"a b c\\">Hello World</button>",
}
`;

exports[`Using the FelaComponent component should compose extension rules 2`] = `
Object {
  "css": ".a {
    font-size: 12px
}

.b {
    color: black
}

.c {
    background-color: blue
}",
  "html": "<button class=\\"a b c\\">Hello World</button>",
}
`;

exports[`Using the FelaComponent component should correctly pass the theme and other props to functional style 1`] = `
Object {
  "css": ".a {
    font-size: 15px
}

.b {
    background-color: blue
}",
  "html": "<div class=\\"a b\\">I am red and written in 15px.</div>",
}
`;

exports[`Using the FelaComponent component should correctly render a fela rule 1`] = `
Object {
  "css": ".a {
    font-size: 12px
}

.b {
    color: red
}",
  "html": "<div class=\\"a b\\">I am red and written in 12px.</div>",
}
`;

exports[`Using the FelaComponent component should default to a div 1`] = `
Object {
  "css": ".a {
    font-size: 12px
}

.b {
    color: red
}",
  "html": "<div class=\\"a b\\"></div>",
}
`;

exports[`Using the FelaComponent component should render children in default mode 1`] = `
Object {
  "css": ".a {
    font-size: 12px
}

.b {
    color: red
}",
  "html": "<div class=\\"a b\\"><span>Hello World</span></div>",
}
`;

exports[`Using the FelaComponent component should render children using the correct as-prop 1`] = `
Object {
  "css": ".a {
    font-size: 12px
}

.b {
    color: red
}",
  "html": "<h1 class=\\"a b\\">Hello World</h1>",
}
`;
